### DISCUSSION TITLE
Changing GAMS solver in message-ix Python API

### DISCUSSION THREAD

**Discussion by Alaa-Eltablawy** [iiasa/message_ix#794] [MESSAGE modeling] [answered]
Hello,

I'm using message_ix Python API on windows 10 and trying to change the default GAMS solver when running scenario.solve()
I've tried different ways by setting gams_agrs, solve_args and solve_options parameters.
Scenario.solve(gams_args=["lp=cbc"])

or 
Scenario.solve(solve_args=["lp=cbc"]

And in both cases, the output in Message_run file shows that LP = CBC is identified, but the GAMS still runs the default solver which is CPLEX

Message_run:
--- Job MESSAGE_run.gms Start ...x86 64bit/MS Windows
--- Applying:
    C:\GAMS\44\gmsprmNT.txt
--- GAMS Parameters defined
    **LP cbc**
    Input C:\..\Lib\site-packages\message_ix\model\MESSAGE_run.gms
    ScrDir C:\..\Lib\site-packages\message_ix\model\225a\
    SysDir C:\GAMS\44\
    LogOption 4
    --in C:\..\Lib\site-packages\message_ix\model\data\MsgData_MODEL_test_baseline.gdx
    --out C:\..\Lib\site-packages\message_ix\model\output\MsgOutput_MODEL_test_baseline.gdx
    --iter C:\..\Lib\site-packages\message_ix\model\output\MsgIterationReport_MODEL_test_baseline.gdx

Is there a way to force the gams_args to overwrite the default gams solver?

**Comment by glatterf42** [comment]
Hi @Alaa-Eltablawy, thanks for reaching out with this question! 
When you call `Scenario.solve(gams_args=["lp=cbc"])`, you are already using the correct parameter in the function. In the end, these options will make their way [here](https://github.com/iiasa/ixmp/blob/049689af2c197a090272842008e1fdd81a080cc9/ixmp/model/gams.py#L166-L188), where both `solve_args` and `gams_args` are set for the model. From [ixmp's documentation](https://docs.messageix.org/projects/ixmp/en/latest/api-model.html#ixmp.model.gams.GAMSModel), we can infer that `gams_args` is probably what you are looking for :)
The documentation also states where you can find the possible options to pass to `gams_args`: https://www.gams.com/latest/docs/UG_GamsCall.html#UG_GamsCall_ListOfCommandLineParameters
My guess is that if you want to change the solver, you want to look at this subsection: https://www.gams.com/latest/docs/UG_GamsCall.html#GAMS_solvers
Unfortunately, this list does not include an option called `cbc`. It does include an option called `LP`, so maybe all you need to do is use all-caps `"LP"="cbc"` rather than `"lp"=...`. Please note that there is also a `solver` option, so if `cbc` is the name of a particular solver, you might want to try `"solver"="cbc"`, too.

**Reply by khaeru** [reply, marked as answer]
If we inspect the MESSAGE GAMS source, we see these lines:
https://github.com/iiasa/message_ix/blob/0fb01b485f8d054c9fe15eda23eee6c07ba2383a/message_ix/model/MESSAGE/auxiliary_settings.gms#L18-L23

@Alaa-Eltablawy, as @glatterf42 says, you do seem to be providing this option to the Python API in a way that reaches the GAMS invocation. However, because this option is **hard-coded** in the source files, the hard-coded value then overwrites the option you've given. (For some other settings, the MESSAGE GAMS source avoids to overwrite a user setting—but not this one.)

The simple fix is to modify the lines I linked above, inserting your chosen LP solver.

**Reply by Alaa-Eltablawy** [reply]
Thanks @khaeru and @glatterf42 for the answers!
Does message_ix initially set the parameters inside  auxilliary_settings.gms based on the GAMS defaults?
So if I'm installing GAMS and setting the default solver to specific solver, and then install the message_ix package, the LP solver inside auxilliary_settings.gms will still be CPLEX or it will be whatever is default in GAMS?

**Reply by glatterf42** [reply]
If I understand correctly, it doesn't matter what you set as default on your system-wide GAMS installation. message_ix will call GAMS with certain options, one of which seems to always be `LP = CPLEX` unless you change the line 23 in `auxiliary_settings.gams`, as @khaeru suggested.
